model/map/src/main/java/org/keycloak/models/map/storage/MapKeycloakTransaction.java

```diff
@@ -17,389 +17,113 @@
-
-    private class MapTaskCompose extends MapTaskWithValue {
-
-        private final MapTaskWithValue oldValue;
-        private final MapTaskWithValue newValue;
-
-        public MapTaskCompose(MapTaskWithValue oldValue, MapTaskWithValue newValue) {
-            super(null);
-            this.oldValue = oldValue;
-            this.newValue = newValue;
-        }
-
-        @Override
-        public void execute() {
-            oldValue.execute();
-            newValue.execute();
-        }
-
-        @Override
-        public V getValue() {
-            return newValue.getValue();
-        }
-
-        @Override
-        public MapOperation getOperation() {
-            return null;
-        }
-
-        @Override
-        public boolean containsCreate() {
-            return oldValue.containsCreate() || newValue.containsCreate();
-        }
-
-        @Override
-        public boolean containsRemove() {
-            return oldValue.containsRemove() || newValue.containsRemove();
-        }
-
-        @Override
-        public boolean isReplace() {
-            return (newValue.getOperation() == MapOperation.CREATE && oldValue.containsRemove()) ||
-              (oldValue instanceof MapKeycloakTransaction.MapTaskCompose && ((MapTaskCompose) oldValue).isReplace());
-        }
-    }
-
-    private class CreateOperation extends MapTaskWithValue {
-        private final K key;
-
-        public CreateOperation(K key, V value) {
-            super(value);
-            this.key = key;
-        }
-
-        @Override public void execute() { map.create(key, getValue()); }
-        @Override public MapOperation getOperation() { return MapOperation.CREATE; }
-    }
-
-    private class UpdateOperation extends MapTaskWithValue {
-        private final K key;
-
-        public UpdateOperation(K key, V value) {
-            super(value);
-            this.key = key;
-        }
-
-        @Override public void execute() { map.update(key, getValue()); }
-        @Override public MapOperation getOperation() { return MapOperation.UPDATE; }
-    }
-
-    private class DeleteOperation extends MapTaskWithValue {
-        private final K key;
-
-        public DeleteOperation(K key) {
-            super(null);
-            this.key = key;
-        }
+    /**
+     * Instructs this transaction to add a new value into the underlying store on commit.
+     *
+     * @param key an identifier that will be associated with the {@code value}
+     * @param value the value
+     */
+    void create(K key, V value);
 
-        @Override public void execute() { map.delete(key); }
-        @Override public MapOperation getOperation() { return MapOperation.DELETE; }
+    /**
+     * Instructs this transaction to update the {@code value} associated with the identifier {@code key} in the
+     * underlying store on commit, when {@code shouldPut} predicate returns true.
+     *
+     * If the underlying store handles transactions on its own, {@code shouldPut} can be ignored
```

Comment:
This method should respect `shouldPut` always to save unnecessary updates.
```suggestion
     * Instructs this transaction to update the {@code value} associated with the identifier {@code key}
     * in the underlying store on commit, if by the time of {@code commit} the {@code shouldPut} predicate 
     * returns {@code true}.
     *
     * If the underlying store handles transactions on its own, {@code shouldPut} can be ignored
```

Reply:
I added this comment because I don't know how to implement this for JPA. Isn't there some logic for not updating when not necessary in JPA already? How we will implement this for JPA?

Reply:
JPA does not implement map transaction or use the Map*Entities - only the new JPA implementation will. So current JPA implementation can be ignored.